### PR TITLE
Cor plot

### PR DESCRIPTION
**Pull request by aegger13**
Fixed up convention for finding fields in .mat file metadata for emit scans
Add cor plot .mat file data object and test
Changed default to need file as arg for object initialization
There is still much to be done with 'beam' metadata.  This might require a whole new object with useful properties or data extraction helper functions.  I believe beam data structure should be the same for both cor plot and emit scans.

Note: We need a cor plot with every measurement checkbox checked to get a solid test file with all available metadata.  This means new properties will need to be added and tests will have to be rewritten/added.

/*start venting Don't ever deal with data like Henrik did, it is not conventional, intuitive, or very useful unless you are him or asked him how to deal with data when he was still around.  I've asked multiple people who use this data about how it is structured and most don't know for sure and/or guess.  This process was infuriating.  End venting.*/

### CODE REVIEW

File: lcls_tools/cor_plot/cor_plot_mat_scan.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        """
+        if CTRL not in self._fields:
+            return None
+
+        idx = self._fields.index(CTRL)
+        ctrls = data[idx]
+        ctrl_list = []
+        for ctrl in ctrls:
+            temp = dict()
+            temp['name'] = str(ctrl[0][0][0])
```
Comment (nneveu) [inline lcls_tools/cor_plot/cor_plot_mat_scan.py]: yikes on this triple 0 index. Just putting a pin in this for later, not sure if there's anything else we can do....

File: lcls_tools/cor_plot/README.MD [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+>>> cpms.beam_names
+('profx', 'xStat', 'xStatStd', 'profy', 'yStat', 'yStatStd', 'profu', 'uStat', 'uStatStd', 'method', 'stats', 'statsStd')
+>>> cpms.timestamp
+737730.05017097027
+>>> cpms.samples
+2
+```
+
+Access Beam Data.  This is very tricky and could use some cleanup.  This is all the metadata associated with the scan and goes by the nature of cpms.beam[iteration][sample][fit].  The fits are in the following order FIT = ['Gaussian', 'Asymmetric', 'Super', 'RMS', 'RMS cut peak', 'RMS cut area', 'RMS floor'].
+
```
Comment (nneveu) [inline lcls_tools/cor_plot/README.MD]: I also want to add a second ping/flag to this structure for cleanup later: cpms.beam[iteration][sample][fit].

File: lcls_tools/cor_plot/README.MD [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+>>> cpms.beam[2][1][4]['profx']
+array([[ -8.04000000e+03,  -8.02392000e+03,  -8.00784000e+03, ...,
+          5.37072000e+03,   5.38680000e+03,   5.40288000e+03],
+       [  0.00000000e+00,   0.00000000e+00,   2.49000000e+02, ...,
+          1.72000000e+02,   0.00000000e+00,   2.64000000e+02],
+       [  1.72791308e+01,   1.77236486e+01,   1.81784282e+01, ...,
+          4.98085984e+00,   4.84116133e+00,   4.70507716e+00]])
+```
+
+As you can see, this whole beam property needs a lot of cleanup.  I do not even know what profx is, maybe an x projection/histogram of the profile monitor?  Your guess is as good as mine.  Good luck!
```
Comment (nneveu) [inline lcls_tools/cor_plot/README.MD]: I agree/think profx is the x projection from the prof monitor.

File: lcls_tools/cor_plot/cor_plot_mat_scan_test.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    'status',
+    'ctrlPV',
+    'beam',
+    'profPV',
+    'ts',
+    'config'
+)
+ACCL_TEST = 'LCLS2'
+CTRL_PV_TEST = 'SOLN:GUNB:212:BCTRL'
+CTRL_VAL_0_TEST = 0.073  # First value of scan ctrl pv
```
Comment (nneveu) [inline lcls_tools/cor_plot/cor_plot_mat_scan_test.py]: Did these constants come from the file?

File: lcls_tools/cor_plot/cor_plot_mat_scan_test.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        self.assertEqual(cpms.iterations, ITER_QE)
+        self.assertEqual(round(cpms.ctrl_vals[0],2), CTRL_VAL_0_QE)
+        self.assertEqual(cpms.beam_names, None)
+        self.assertEqual(round(cpms.timestamp, 2), TS_QE)
+
+    def test_init_bad_file(self):
+        self.assertRaises(Exception, CPMS(BAD_FILE))
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (nneveu) [inline lcls_tools/cor_plot/cor_plot_mat_scan_test.py]: GitHub wants a new line?
```suggestion
    unittest.main()
    
```

File: lcls_tools/emit_scan/mat_emit_scan.py [hunk 6]
```diff
@@ -11,196 +14,247 @@
...
+    def mat_file(self):
+        """Mat file loaded"""
         return self._file
 
     @property
     def status(self):
-        return self._status
+        """Array of 0 or 1, no idea what status it's looking at"""
+        if self._status is not None:
+            return [status[0] for status in self._status]
```
Comment (nneveu) [inline lcls_tools/emit_scan/mat_emit_scan.py]: Some of the zero indexing here on out is scary again....but forced by Henrik.
Flagging for later.

File: lcls_tools/cor_plot/cor_plot_mat_scan.py (hunk 1 above)
Comment (aegger13) [reply lcls_tools/cor_plot/cor_plot_mat_scan.py]: Yeah, Matlab structures come through as giant nested lists (if only he had used Matlab dictionaries).  What could be done is if someone revamped the Matlab data saving code, but it's fairly hideous... poor Tim

File: lcls_tools/cor_plot/cor_plot_mat_scan_test.py (hunk 4 above)
Comment (aegger13) [reply lcls_tools/cor_plot/cor_plot_mat_scan_test.py]: They do come from the files, it's a way of verifying that we always get the same answer based on any code changes

File: lcls_tools/cor_plot/cor_plot_mat_scan_test.py (hunk 5 above)
Comment (aegger13) [reply lcls_tools/cor_plot/cor_plot_mat_scan_test.py]: Hmmm... not sure if that's a PEP8 thing, I know I've seen this sort of suggestion before, not sure why

**Pull request merged**